### PR TITLE
Clarify SETTINGS, 0-RTT, HTTP/3, etc.

### LINKED ISSUE #2: Is SETTINGS frame on HTTP/3 control stream allowed if there is a SETTINGS frame in ALPS?
Section 3, "Use of ALPS in HTTP", makes a SETTINGS frame in ALPS "supersede the requirement to send a SETTINGS frame at the beginning of the HTTP/2 connection", which makes perfect sense.  The way I interpret "superseding the requirement" is that it is not required, but there is no prohibition on sending further SETTINGS frames on stream 0, which is fine for HTTP/2.

However, in case of HTTP/3, a SETTINGS frame in ALPS "supersedes the requirement to send a SETTINGS frame at the beginning of the control stream".  It does not explicitly ban a SETTINGS frame as the first frame on the control stream, and does not explicitly mention if the ban on sending a SETTINGS frame as the non-first frame on the control stream is affected.  I propose that we spell out explicitly that in case a SETTINGS frame in sent in ALPS, no SETTINGS frames MUST be sent on the control stream.  This is in alignment with HTTP/3's requirement that there must be exactly one SETTINGS frame.

### LINKED ISSUE #4: Guidance on 0-RTT and varying fields
ALPS's 0-RTT integration assumes that anything you put in ALPS is mostly-constant and wants to be preserved over 0-RTT. It's okay for the value to change, but 0-RTT will miss, so the expectation is it doesn't happen often. (E.g. server code or config change.)

This is somewhat a fundamental trade-off: if you want reliable ordering and 0-RTT, you to be okay with tying 0-RTT to the value not changing. But different SETTINGS values may have different preferences here. [draft-thomson-httpbis-h2-0rtt-00](https://tools.ietf.org/html/draft-thomson-httpbis-h2-0rtt-00)'s approach is to have every value either be remembered or not in the registry, but this involves some non-trivial h2/TLS integrations.

My preference is we simply say this is part of the properties ALPS provides (just like its relationship with client certs is one of those properties), and protocols use it accordingly. So the h2 and h3 mapping could say callers send their fixed, authentication-not-required SETTINGS in ALPS and varying SETTINGS outside ALPS. If we go with that, we should have text to that effect.

### PR DESCRIPTION
**Pull request by davidben**
@bencebeky @vasilvv, here's a pass at some text. How does this look? The TLS half of the text is https://github.com/vasilvv/tls-alps/pull/10.

Closes #2. Closes #4.

### CODE REVIEW

File: draft-vvv-httpbis-alps.md [hunk 1]
```diff
@@ -85,38 +85,102 @@ when, and only when, they appear in all capitals, as shown here.
...
+([HTTP3], Section 7.2.4).  Frames defined in later documents may be allowed in
+ALPS, so the receiver MUST ignore unrecognized frames in the ALPS payload.
+
+If ALPS is successfully negotiated during a TLS handshake for an HTTP/2
+connection, the protocol is updated as follows:
+
+ALPS updates the initial value of each setting as described in Section 7.2.4.2
+of [HTTP3]. In both the client and server, the initial value of each setting
+is the value in the ALPS SETTINGS frame, or the default value if not specified.
+This allows messages to incoporate non-default settings values without waiting
```
Comment (bencebeky) [inline draft-vvv-httpbis-alps.md]: s/incoporate/incorporate/

File: draft-vvv-httpbis-alps.md [hunk 2]
```diff
@@ -85,38 +85,102 @@ when, and only when, they appear in all capitals, as shown here.
...
+empty, or it MAY contain additional settings advertised outside of ALPS. See
+{{early-data}} for guidance. This frame MUST NOT reduce any limits or alter any
+values that might be violated by a peer using the initial values from ALPS. If
+the peer sends an incompatible SETTINGS frame, the receiver MUST treat it as a
+connection error of type H3_SETTINGS_ERROR.
+
+ALPS additionally replaces the 0-RTT mechanisms in Section 7.4.2.4 of [HTTP3].
+Clients and servers SHOULD NOT separately remember settings with the TLS
+session. Instead, clients attempting 0-RTT MUST comply with the initial values
+of each setting, as determined above. Note these values will incorporate the
```
Comment (bencebeky) [inline draft-vvv-httpbis-alps.md]: s/Note/Note that/

File: draft-vvv-httpbis-alps.md [hunk 3]
```diff
@@ -85,38 +85,102 @@ when, and only when, they appear in all capitals, as shown here.
...
+Since settings exchanged through ALPS are always available at the beginning of
+the connection, HTTP/3 extensions MAY opt to require those to be sent through
+ALPS.
+
+[[OPEN ISSUE: Is SETTINGS in ALPS mandatory or optional? Should it be the
+first frame?]]
+
+# Interaction with Early Data {#early-data}
+
+ALPS introduces two ways to send HTTP/2 and HTTP/3 settings values: inside
```
Comment (bencebeky) [inline draft-vvv-httpbis-alps.md]: Well, technically one of the ways is already there, ALPS only adds a second one.  Maybe this sentence could be reworded?

File: draft-vvv-httpbis-alps.md (hunk 1 above)
Comment (bencebeky) [inline draft-vvv-httpbis-alps.md]: s/settings values/setting values/, here and twice more below.

See https://httpwg.org/specs/rfc7540.html#SETTINGS and https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-settings for terminology: a SETTINGS frame carries parameters known as settings, each setting (singular) has a setting value and a setting identifier.

File: draft-vvv-httpbis-alps.md [hunk 4]
```diff
@@ -85,38 +85,102 @@ when, and only when, they appear in all capitals, as shown here.
...
+that change on a per-connection basis will reduce 0-RTT acceptance.
+
+In contrast, values sent over the control stream can change over time without
+affected 0-RTT. However, they are not available at the start of the connection
+and may not be applied to all messages in a connection, notably those sent over
+0-RTT.
+
+It is expected that most setting values reflect implementation configuration or
+capabilities, which do not change frequently. These values SHOULD be sent
+inside ALPS. However, mechanisms like GREASE {{?RFC8701}} that randomize values
```
Comment (bencebeky) [inline draft-vvv-httpbis-alps.md]: Maybe reference reserved setting identifiers defined in core HTTP/3 at https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-defined-settings-parameters.  Note that HTTP/3 does not call it GREASE.

Unfortunately the HTTP/2 spec does not incorporate SETTINGS GREASE.  Not sure if it's appropriate to reference Mike Bishop's draft: https://mikebishop.github.io/http-misc-extensions/draft-bishop-httpbis-grease.html which has expired, not sure if it matters.

File: draft-vvv-httpbis-alps.md (hunk 1 above)
Comment (davidben) [reply draft-vvv-httpbis-alps.md]: Done.

File: draft-vvv-httpbis-alps.md (hunk 4 above)
Comment (davidben) [reply draft-vvv-httpbis-alps.md]: Done.

File: draft-vvv-httpbis-alps.md (hunk 3 above)
Comment (davidben) [reply draft-vvv-httpbis-alps.md]: Done.

File: draft-vvv-httpbis-alps.md (hunk 2 above)
Comment (davidben) [reply draft-vvv-httpbis-alps.md]: Done.

File: draft-vvv-httpbis-alps.md (hunk 1 above)
Comment (davidben) [reply draft-vvv-httpbis-alps.md]: Done.

**Comment by davidben** [comment]
> I think it is customary to write "on the control stream" instead of "in the control stream"

Done.

**Review by bencebeky — APPROVED** [review]

**Pull request merged**